driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -458,8 +484,56 @@ public void close() {
                 if (LOGGER.isTraceEnabled()) {
                     LOGGER.trace(format("Checked in connection [%s] to server %s", getId(wrapped), serverId.getAddress()));
                 }
-                pool.release(wrapped, wrapped.isClosed() || shouldPrune(wrapped));
+                boolean success = false;
+                try {
+                    if (wrapped.isClosed() || shouldPrune(wrapped)) {
+                        pool.release(wrapped, true);
+                    } else if (!openConcurrencyLimiter.tryHandOver(new PooledConnection(wrapped))) {
+                        pool.release(wrapped);
+                    }
+                    success = true;
+                } finally {
+                    if (!success) {
```

Comment:
We have `wrapped.isClosed()`, `shouldPrune(wrapped)` and `tryHandOver` also called here. `tryHandOver` definitely may throw an unchecked exception if the thread detects it was interrupted while trying to acquire the lock. Btw, the same may be true for many of our methods that use interruptible Java SE API, but do not re-throw `InterruptedException`. Other unchecked exceptions (including those wrapping checked exceptions) also may happen where we do not expect.

Even if it does not make a difference here (I am guessing, that if the thread is interrupted then the client is in the process of being closed), I do not see how robust code may be written by specifying "Must not throw Exceptions" on many methods and then relying on that instead of using try-finally and try-with-resources blocks. This is not a straightforward question.

Anyway, I removed the guarding code here and in other places.

Reply:
After thinking more, I realized the following with regard to interruptions:
1. Both `releasePermit` and `tryHandOver` methods acquire lock uninterruptibly (that's just how I wrote the code). `releasePermit` must do so, while `tryHandOver` may use `Lock.lockInterruptibly`.
2. I decided to leave both `releasePermit` and `tryHandOver` as is (`tryHandOver` cannot be interrupted), but made it more clear in the code that they use the `Lock.lock` method.
3. My statement _"if the thread is interrupted then the client is in the process of being closed"_ is incorrect: for synchronous driver API this thread is most likely a user thread, because `com.mongodb.client.MongoClient` does not maintain a pool of worker threads. This means that if we are not careful, an interrupted user thread may leave a `MongoClient` in a state that violates internal assumptions, thus rendering the client unavailable even though a user simply interrupted his own thread. To me this is a revelation to some extent, but it may be obvious to others, and it definitely does not mean that we have problems like that in the current code. However, if I were to use `lockInterruptibly` in `tryHandOver` (which I could, and for some reason I incorrectly assumed I did), then removing a try-finally block in `PooledConnection.close` would result in us potentially leaking connections.

I made a small change f5967d1ae9f177a59c55a92fbdc6a52ca51f5f27 to make it clearer in the code that `releasePermit` and `tryHandOver` use uninterruptible API.